File: leetcode/20.cpp
#include <iostream>
#include <string>
#include <stack>

using namespace std;

class Solution {
public:
  bool isValid(string s) {
    stack<char> help;
    for (string::iterator it = s.begin(); it != s.end(); ++it) {
      if (help.empty()) {
        help.push(*it);
      } else if ((help.top() == '(' && (*it) == ')') ||
                 (help.top() == '{' && (*it) == '}') ||
                 (help.top() == '[' && (*it) == ']')) {
        help.pop();
      } else {
        help.push(*it);
      }
    }
    return help.empty();
  }
};

File: leetcode/20.py
class Solution:

    def isValid(self, s: str) -> bool:
        stack = []

        for c in s:
            if len(stack) == 0:
                stack.append(c)
            else:
                if (c == ")" and stack[-1] == "(") or (c == "]" and stack[-1] == "[") or (c == "}" and stack[-1] == "{"):
                    stack.pop()
                else:
                    stack.append(c)

        return len(stack) == 0